cellbase-lib/src/main/java/org/opencb/cellbase/lib/impl/core/DataReleaseSingleton.java

```diff
@@ -0,0 +1,199 @@
+    private Map<String, ReentrantReadWriteLock> rwLockMap = new HashMap<>();
+
+    // {key = dbname, value = { key = release, value = { key = data, value = collection } } }
+    private Map<String, Map<Integer, Map<String, MongoDBCollection>>> cachedData = new HashMap<>();
+
+    private CellBaseManagerFactory managerFactory;
+
+    private static DataReleaseSingleton instance;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DataReleaseSingleton.class);
+
+    public static final String UNKOWN_DATABASE_MSG_PREFIX = "Unknown database ";
+    public static final String INVALID_RELEASE_MSG_PREFIX = "Invalid release ";
+    public static final String INVALID_DATA_MSG_PREFIX = "Invalid data ";
+
+    // Private constructor to prevent instantiation
+    private DataReleaseSingleton(CellBaseManagerFactory managerFactory) throws CellBaseException {
+        this.managerFactory = managerFactory;
+
+        // Support multi species and assemblies
+        CellBaseConfiguration configuration = managerFactory.getConfiguration();
+        for (SpeciesConfiguration vertebrate : configuration.getSpecies().getVertebrates()) {
+            for (SpeciesConfiguration.Assembly assembly : vertebrate.getAssemblies()) {
+                String databaseName = MongoDBManager.getDatabaseName(vertebrate.getId(), assembly.getName(), configuration.getVersion());
+                // This is necessary, before creating the database name the assembly is "cleaned", and we need to get the data release
+                // manager from the species and the assembly
+                speciesMap.put(databaseName, vertebrate.getId());
+                assemblyMap.put(databaseName, assembly.getName());
+                rwLockMap.put(databaseName, new ReentrantReadWriteLock());
+                cachedData.put(databaseName, new HashMap<>());
+
+                MongoClient mongoClient = managerFactory.getDataReleaseManager(vertebrate.getId(), assembly.getName()).getMongoDatastore()
+                        .getMongoClient();
+                MongoDatabase database = mongoClient.getDatabase(databaseName);
+                MongoCollection<Document> collection = database.getCollection(ReleaseMongoDBAdaptor.DATA_RELEASE_COLLECTION_NAME);
+                LOGGER.info("Setting listener for database {} and collection {}", database.getName(), collection.getNamespace()
+                        .getCollectionName());
+                // Set up the change stream for the collection
+                new Thread(() -> {
+                    collection.watch().fullDocument(FullDocument.UPDATE_LOOKUP).forEach(changeStreamDocument -> {
+                        try {
+                            handleDocumentChange(changeStreamDocument);
+                        } catch (CellBaseException e) {
+                            LOGGER.warn("Exception from handle document change function: {}", e.getStackTrace());
+                        }
+                    });
+                }).start();
```

Comment:
All these threads should be within a single "ExecutionService". Then, as this is a singleton class, and can't be closed easily, a ShutdownHook should terminate the ExecutionService
```java
    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
        threadPool.shutdownNow();
    });
```